### DISCUSSION TITLE
Existence Check - inverse of Inverted Section

### DISCUSSION THREAD

**Discussion by clach04** [mustache/spec#171] [Ideas]
Option 3 in https://github.com/mustache/spec/discussions/157 in Stache for Python.

Copy/paste extract from readme:

### Existence Check {{?exists}}{{/}}

See https://github.com/mustache/spec/discussions/157

Forces a check of the tag name, rather than imply that it is a section block. This
is useful for check if an array has members rather than iterate over the members

    {{?array}}
    stuff\n
    {{/}}

with `{array: [1, 2, 3, 4]}` results in:

    stuff

as opposed to

    {{#array}}
    stuff\n
    {{/}}

which would render

    stuff
    stuff
    stuff
    stuff

I can't claim any ownership of this, this was implemented in https://github.com/hyperturtle/Stache by hyperturtle, more about my fork https://github.com/mustache/spec/discussions/172

**Comment by clach04** [comment]
I make extensive use of this, and I've seen a number of requests (see https://github.com/mustache/spec/discussions/157#discussioncomment-7541174 for links to some requests) for this functionality (even if not this syntax). Typically used to avoid including html content for a section or even boiler plate for a table so as to avoid an empty table body. My primary use case is omitting entire sections, not just the table boilerplate.

Whilst I like this syntax, I'm not wedded to it. I could happily use an alternative syntax based on census in https://github.com/mustache/spec/discussions/157

**Comment by determin1st** [comment]
```
[[^array]]
  is empty
[[|]]
  is not empty
[[/]]

[[#array]]
  [[.]]
[[|]]
  is empty
[[/]]
```

**Reply by jgonggrijp** [reply]
That is also an interesting solution I had not thought of. Shall I add it to the table in #157?

**Reply by determin1st** [reply]
as you wish

**Reply by jgonggrijp** [reply]
Done.

**Comment by clach04** [comment]
Here are 3 uses that I've cobbled together from some real use cases I have at work:

## variables

    {
        "code": "void main()\n{\n}\n",
        "crimes": [
            "like country music",
            "leave the toliet seat up"
        ],
        "crime_details": [
            {
                "crime": "like country music",
                "url": "http://somewhere/1"
            },
            {
                "crime": "leave the toliet seat up",
                "url": "http://somewhere/2"
            }
        ]
    }

## Template

    <h1>Demo not exists use cases</h1>
    
    <h2>Code block tags</h2>
    
    {{?code}}
    <pre><code>
    {{code}}
    </code></pre>
    {{/code}}
    
    <h2>Simple Value Lists(i.e. not hash/dictionary)</h2>
    
    {{! Only show crimes text and list if there is one. NOTE Stache only feature }}
    
    {{?crimes}}
    <p>You are a criminal and a generally awful person in general. These are the crimes you are guilty of:
    </p>
    
    <ol>
    {{/crimes}}
    
    {{#crimes}}
        <li>{{.}}</li>
    {{/crimes}}
    
    {{?crimes}}
    </ol>
    {{/crimes}}
    
    <h2>Hash/Dictionary Lists</h2>
    
    {{! Only show crimes text and list if there is one. NOTE Stache only feature }}
    
    {{?crime_details}}
    <p><p>You are a criminal and a generally awful person in general. These are the crimes you are guilty of:
    </p>
    
    <ol>
    </p>
    {{/crime_details}}
    
    {{#crime_details}}
        <li><a href="{{{url}}}">{{crime}}</a></li>
    {{/crime_details}}
    
    {{?crimes}}
    </ol>
    {{/crimes}}
    
## Rendered Output

    <h1>Demo not exists use cases</h1>
    
    <h2>Code block tags</h2>
    
    <pre><code>
    void main()
    {
    }
    
    </code></pre>
    
    <h2>Simple Value Lists(i.e. not hash/dictionary)</h2>
    
    
    <p>You are a criminal and a generally awful person in general. These are the crimes you are guilty of:
    </p>
    
    <ol>
    
        <li>like country music</li>
        <li>leave the toliet seat up</li>
    
    </ol>
    
    <h2>Hash/Dictionary Lists</h2>
    
    
    <p><p>You are a criminal and a generally awful person in general. These are the crimes you are guilty of:
    </p>
    
    <ol>
    </p>
    
        <li><a href="http://somewhere/1">like country music</a></li>
        <li><a href="http://somewhere/2">leave the toliet seat up</a></li>
    
    </ol>
    
    
    
    <h1>Demo not exists use cases</h1>
    
    <h2>Code block tags</h2>
    
    <pre><code>
    void main()
    {
    }
    
    </code></pre>
    
    <h2>Simple Value Lists(i.e. not hash/dictionary)</h2>
    
    
    <p>You are a criminal and a generally awful person in general. These are the crimes you are guilty of:
    </p>
    
    <ol>
    
        <li>like country music</li>
        <li>leave the toliet seat up</li>
    
    </ol>
    
    <h2>Hash/Dictionary Lists</h2>
    
    
    <p><p>You are a criminal and a generally awful person in general. These are the crimes you are guilty of:
    </p>
    
    <ol>
    </p>
    
        <li><a href="http://somewhere/1">like country music</a></li>
        <li><a href="http://somewhere/2">leave the toliet seat up</a></li>
    
    </ol>

I do like the `^ |` syntax BUT for my selfish use cases (above), I'd have to use something like:

    {{^variable}}{{|}}
    stuff here
    {{/variable}}

I'd rather the proposed `^^` instead, but again I do like the  `^ |` syntax and would likely have some uses cases for it.

I feel more experienced with Mustache when I first wrote that code sample above so I'd likely not write a template in that style BUT for the list examples I do need a plain not exists (not falsey) check.

**Reply by jgonggrijp** [reply]
The `code` example would work with a regular section, but otherwise, I agree there is a need for something like a list nonemptiness check. Your examples are pretty typical and unsurprising for this use case. I don't mean that as a bad thing, by the way.

While there are some people who think no new solution is required, I expect most of the controversy to be about *how* to solve it if you do. We already have #157 for that discussion, so I will not go into it very deeply over here.

My objection to the current proposal, a dedicated sigil `{{?`, is that we would be introducing a change to the language just for this single use case. `{{^^` has the same problem, except that it also has far-reaching implications for the grammar of the language: sigils suddenly behaving like unary operators that can be combined OR sigils suddenly being multi-character.

*If* we introduce some kind of new feature in Mustache that solves the list nonemptiness problem (and again, I agree we should), then please let it be something that *also* solves a whole range of other problems. @determin1st's `{{|}}` already does much better in that regard, but I personally still prefer filters.

**Comment by determin1st** [comment]
```
{{?crimes}}
  <ol>
  {{#crimes}}
    <li>{{.}}</li>
  {{/}}
  </ol>
{{/}}
```

VS

```
{{^crimes}}
{{|}}
  <ol>
  {{#crimes}}
    <li>{{.}}</li>
  {{/}}
  </ol>
{{/}}
```

VS

```
{{#crimes-boolean-flag-maybe-from-the-helper-you-push-along-the-data}}
  <ol>
  {{#crimes}}
    <li>{{.}}</li>
  {{/}}
  </ol>
{{/}}
```

imo, the third is the best one.
using lambdas with auto-ejecting helpers allows to split big templates into parts, for example

```
...
{{put crimes}}
...
```

where `put` is a lambda (a function) which will check whether crimes is non-empty and optionally create and push `crimes` helper onto the stack

**Reply by jgonggrijp** [reply]
What are "auto-ejecting helpers"?

**Reply by determin1st** [reply]
those helpers that are pushed by lambda (inside the function's body).

ejection procedure must differ for a variable and a block. this improvement to lambdas will allow cleaner solution where previously data enrichment/mixin took place.
